File: frontend/types/trend-analysis.ts
/**
 * Trend Analysis Types
 * TypeScript types for the AI trend analysis feature.
 */

// ============================================
// ENUMS
// ============================================

export type TrendDirection = 'rising' | 'falling' | 'stable' | 'volatile';

export type TrendCategory =
  | 'viral_positive'
  | 'viral_negative'
  | 'competitor_launch'
  | 'seasonal'
  | 'news_event'
  | 'market_shift'
  | 'organic_growth'
  | 'organic_decline';

export type OpportunityType =
  | 'price_increase'
  | 'price_decrease'
  | 'hold'
  | 'promotional'
  | 'premium_positioning';

export type RiskLevel = 'low' | 'medium' | 'high' | 'critical';

export type ConfidenceLevel = 'low' | 'medium' | 'high' | 'very_high';

// ============================================
// REQUEST TYPES
// ============================================

export interface TrendAnalysisRequest {
  days?: number;
  product_ids?: string[];
  use_model?: 'openai' | 'gemini';
}

export interface ProductOpportunityRequest {
  product_id: string;
  use_model?: 'openai' | 'gemini';
}

export interface RiskDetectionRequest {
  use_model?: 'openai' | 'gemini';
}

export interface InsightGenerationRequest {
  days?: number;
  use_model?: 'openai' | 'gemini';
}

// ============================================
// RESPONSE TYPES
// ============================================

export interface TrendSignal {
  signal_type: string;
  value: number;
  timestamp: string;
  source: string;
  description: string;
}

export interface TrendPrediction {
  direction: TrendDirection;
  category: TrendCategory;
  confidence: ConfidenceLevel;
  confidence_score: number;
  predicted_change: number;
  timeframe_days: number;
  reasoning: string;
  supporting_signals: TrendSignal[];
}

export interface PricingOpportunity {
  opportunity_type: OpportunityType;
  product_id: string;
  product_name: string;
  current_price: string;
  suggested_price: string;
  expected_impact: string;
  confidence: ConfidenceLevel;
  confidence_score: number;
  reasoning: string;
  valid_until: string;
  triggers: string[];
}

export interface RiskAlert {
  risk_level: RiskLevel;
  risk_type: string;
  title: string;
  description: string;
  affected_products: string[];
  recommended_actions: string[];
  detected_at: string;
  expires_at: string | null;
}

export interface AIInsight {
  title: string;
  summary: string;
  detailed_analysis: string;
  key_factors: string[];
  data_points_analyzed: number;
  generated_at: string;
  model_used: string;
}

export interface TrendAnalysisResponse {
  analysis_id: string;
  generated_at: string;
  market_sentiment: TrendDirection;
  market_sentiment_score: number;
  predictions: TrendPrediction[];
  opportunities: PricingOpportunity[];
  risks: RiskAlert[];
  insights: AIInsight[];
  executive_summary: string;
  recommended_actions: string[];
  products_analyzed: number;
  mentions_analyzed: number;
  time_range_days: number;
}

export interface RiskDetectionResponse {
  risks: RiskAlert[];
  overall_risk_level: RiskLevel;
  summary: string;
  generated_at: string;
}

export interface QuickStatsResponse {
  current_sentiment: number;
  sentiment_trend: TrendDirection;
  sentiment_change_7d: number;
  mentions_today: number;
  mentions_7d: number;
  volume_change_percent: number;
  active_opportunities: number;
  potential_revenue_impact: string;
  active_risks: number;
  highest_risk_level: RiskLevel;
  trending_up: string[];
  trending_down: string[];
  last_updated: string;
}

// ============================================
// UI HELPER TYPES
// ============================================

export interface TrendDisplayInfo {
  label: string;
  color: string;
  icon: string;
  bgColor: string;
}

export interface RiskDisplayInfo {
  label: string;
  color: string;
  bgColor: string;
  borderColor: string;
}

export interface ConfidenceDisplayInfo {
  label: string;
  color: string;
  width: string;
}

// ============================================
// HELPER FUNCTIONS
// ============================================

export function getTrendDisplayInfo(direction: TrendDirection): TrendDisplayInfo {
  const info: Record<TrendDirection, TrendDisplayInfo> = {
    rising: {
      label: 'Rising',
      color: 'text-green-600',
      icon: '↑',
      bgColor: 'bg-green-100',
    },
    falling: {
      label: 'Falling',
      color: 'text-red-600',
      icon: '↓',
      bgColor: 'bg-red-100',
    },
    stable: {
      label: 'Stable',
      color: 'text-gray-600',
      icon: '→',
      bgColor: 'bg-gray-100',
    },
    volatile: {
      label: 'Volatile',
      color: 'text-orange-600',
      icon: '↕',
      bgColor: 'bg-orange-100',
    },
  };
  return info[direction];
}

export function getRiskDisplayInfo(level: RiskLevel): RiskDisplayInfo {
  const info: Record<RiskLevel, RiskDisplayInfo> = {
    low: {
      label: 'Low',
      color: 'text-green-700',
      bgColor: 'bg-green-100',
      borderColor: 'border-green-300',
    },
    medium: {
      label: 'Medium',
      color: 'text-yellow-700',
      bgColor: 'bg-yellow-100',
      borderColor: 'border-yellow-300',
    },
    high: {
      label: 'High',
      color: 'text-orange-700',
      bgColor: 'bg-orange-100',
      borderColor: 'border-orange-300',
    },
    critical: {
      label: 'Critical',
      color: 'text-red-700',
      bgColor: 'bg-red-100',
      borderColor: 'border-red-300',
    },
  };
  return info[level];
}

export function getConfidenceDisplayInfo(level: ConfidenceLevel): ConfidenceDisplayInfo {
  const info: Record<ConfidenceLevel, ConfidenceDisplayInfo> = {
    low: {
      label: 'Low Confidence',
      color: 'bg-gray-400',
      width: 'w-1/4',
    },
    medium: {
      label: 'Medium Confidence',
      color: 'bg-yellow-500',
      width: 'w-1/2',
    },
    high: {
      label: 'High Confidence',
      color: 'bg-green-500',
      width: 'w-3/4',
    },
    very_high: {
      label: 'Very High Confidence',
      color: 'bg-green-600',
      width: 'w-full',
    },
  };
  return info[level];
}

export function getOpportunityTypeLabel(type: OpportunityType): string {
  const labels: Record<OpportunityType, string> = {
    price_increase: 'Price Increase',
    price_decrease: 'Price Decrease',
    hold: 'Hold Current Price',
    promotional: 'Run Promotion',
    premium_positioning: 'Premium Positioning',
  };
  return labels[type];
}

export function getCategoryLabel(category: TrendCategory): string {
  const labels: Record<TrendCategory, string> = {
    viral_positive: 'Viral (Positive)',
    viral_negative: 'Viral (Negative)',
    competitor_launch: 'Competitor Launch',
    seasonal: 'Seasonal Trend',
    news_event: 'News Event',
    market_shift: 'Market Shift',
    organic_growth: 'Organic Growth',
    organic_decline: 'Organic Decline',
  };
  return labels[category];
}

export function formatSentimentScore(score: number): string {
  if (score > 0) return `+${score.toFixed(1)}`;
  return score.toFixed(1);
}

export function formatPercentChange(change: number): string {
  const sign = change >= 0 ? '+' : '';
  return `${sign}${change.toFixed(1)}%`;
}

File: backend/schemas/trend_analysis.py
"""
Trend Analysis API Schemas

Pydantic schemas for the trend analysis API endpoints.
"""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

# ============================================
# ENUMS
# ============================================


class TrendDirection(StrEnum):
    RISING = "rising"
    FALLING = "falling"
    STABLE = "stable"
    VOLATILE = "volatile"


class TrendCategory(StrEnum):
    VIRAL_POSITIVE = "viral_positive"
    VIRAL_NEGATIVE = "viral_negative"
    COMPETITOR_LAUNCH = "competitor_launch"
    SEASONAL = "seasonal"
    NEWS_EVENT = "news_event"
    MARKET_SHIFT = "market_shift"
    ORGANIC_GROWTH = "organic_growth"
    ORGANIC_DECLINE = "organic_decline"


class OpportunityType(StrEnum):
    PRICE_INCREASE = "price_increase"
    PRICE_DECREASE = "price_decrease"
    HOLD = "hold"
    PROMOTIONAL = "promotional"
    PREMIUM_POSITIONING = "premium_positioning"


class RiskLevel(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ConfidenceLevel(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    VERY_HIGH = "very_high"


# ============================================
# REQUEST SCHEMAS
# ============================================


class TrendAnalysisRequest(BaseModel):
    """Request to run AI trend analysis."""

    days: int = Field(default=30, ge=7, le=90, description="Number of days to analyze (7-90)")
    product_ids: list[str] | None = Field(
        default=None, description="Specific product IDs to analyze (None = all products)"
    )
    use_model: str = Field(
        default="openai", pattern="^(openai|gemini)$", description="AI model to use: 'openai' or 'gemini'"
    )


class ProductOpportunityRequest(BaseModel):
    """Request to analyze a specific product for opportunities."""

    product_id: str = Field(..., description="Product ID to analyze")
    use_model: str = Field(default="openai", pattern="^(openai|gemini)$", description="AI model to use")


class RiskDetectionRequest(BaseModel):
    """Request to detect risks."""

    use_model: str = Field(default="openai", pattern="^(openai|gemini)$", description="AI model to use")


class InsightGenerationRequest(BaseModel):
    """Request to generate market insight."""

    days: int = Field(default=30, ge=7, le=90, description="Number of days to analyze")
    use_model: str = Field(default="openai", pattern="^(openai|gemini)$", description="AI model to use")


# ============================================
# RESPONSE SCHEMAS
# ============================================


class TrendSignalResponse(BaseModel):
    """A single trend signal."""

    signal_type: str
    value: float
    timestamp: datetime
    source: str
    description: str


class TrendPredictionResponse(BaseModel):
    """A trend prediction."""

    direction: TrendDirection
    category: TrendCategory
    confidence: ConfidenceLevel
    confidence_score: float = Field(ge=0, le=100)
    predicted_change: float
    timeframe_days: int
    reasoning: str
    supporting_signals: list[TrendSignalResponse] = []


class PricingOpportunityResponse(BaseModel):
    """A pricing opportunity."""

    opportunity_type: OpportunityType
    product_id: str
    product_name: str
    current_price: str
    suggested_price: str
    expected_impact: str
    confidence: ConfidenceLevel
    confidence_score: float = Field(ge=0, le=100)
    reasoning: str
    valid_until: datetime
    triggers: list[str] = []


class RiskAlertResponse(BaseModel):
    """A risk alert."""

    risk_level: RiskLevel
    risk_type: str
    title: str
    description: str
    affected_products: list[str] = []
    recommended_actions: list[str] = []
    detected_at: datetime
    expires_at: datetime | None = None


class AIInsightResponse(BaseModel):
    """An AI-generated insight."""

    title: str
    summary: str
    detailed_analysis: str
    key_factors: list[str] = []
    data_points_analyzed: int
    generated_at: datetime
    model_used: str


class TrendAnalysisResponse(BaseModel):
    """Complete trend analysis result."""

    model_config = ConfigDict(from_attributes=True)
    analysis_id: str
    generated_at: datetime

    # Market overview
    market_sentiment: TrendDirection
    market_sentiment_score: float = Field(ge=-100, le=100)

    # Analysis results
    predictions: list[TrendPredictionResponse] = []
    opportunities: list[PricingOpportunityResponse] = []
    risks: list[RiskAlertResponse] = []
    insights: list[AIInsightResponse] = []

    # Summary
    executive_summary: str
    recommended_actions: list[str] = []

    # Metadata
    products_analyzed: int
    mentions_analyzed: int
    time_range_days: int


class RiskDetectionResponse(BaseModel):
    """Risk detection result."""

    risks: list[RiskAlertResponse] = []
    overall_risk_level: RiskLevel
    summary: str
    generated_at: datetime


class QuickStatsResponse(BaseModel):
    """Quick stats for the trends dashboard."""

    # Sentiment
    current_sentiment: float
    sentiment_trend: TrendDirection
    sentiment_change_7d: float

    # Volume
    mentions_today: int
    mentions_7d: int
    volume_change_percent: float

    # Opportunities
    active_opportunities: int
    potential_revenue_impact: str

    # Risks
    active_risks: int
    highest_risk_level: RiskLevel

    # Products
    trending_up: list[str]
    trending_down: list[str]

    last_updated: datetime
